### ISSUE TITLE
Erroneous Connection does not fail

### ISSUE THREAD

**Issue by bongolegend** [confluentinc/confluent-kafka-python#1048]
Description
===========

When I initialize a new consumer, if I pass an erroneous connection string via `bootstrap.servers`, I don't get an error. I would expect to get an error; if not at initialization, then upon `poll`.

How to reproduce
================

confluent-kafka == ('1.4.2', 17039872)
librdkafka ==  ('1.4.2', 17040127)

MacOS 11.1 (big sur)

```
import confluent_kafka as ck
config = {'bootstrap.servers': 'fake', 'group.id': 'fake'}
con = ck.Consumer(**config)
con.subscribe(['nonexistant-topic'])
con.poll(timeout=0.1)
```

Checklist
=========
Please provide the following information:

 - [ x] confluent-kafka-python and librdkafka version (`confluent_kafka.version()` and `confluent_kafka.libversion()`):
 - [ ] Apache Kafka broker version:
 - [ x] Client configuration: `{...}`
 - [ x] Operating system:
 - [ ] Provide client logs (with `'debug': '..'` as necessary)
 - [ ] Provide broker log excerpts
 - [ ] Critical issue

**Comment by edenhill** [comment]
Add an error_cb (callable) to your config and look for the KafkaError ALL_BROKERS_DOWN.
Example here:
https://github.com/confluentinc/confluent-kafka-python/blob/77681916850ec5902dd0b9f270f8274d73d81b25/tests/test_KafkaError.py#L10

**Comment by edenhill** [comment]
Authentication errors are also propagated through the error_cb, with error code set to `_AUTH`.

Do note that the client constructor is asynchronous and will not return/raise an error if there is a problem connecting to the brokers, that happens later and thus the use of the error_cb.

**Comment by ncernek** [comment]
I expect to catch authentication errors in the example above, as I am using `error_cb`, and telling it to raise on any error. 

>  that happens later

What do you mean by "later" ? When attempting a `poll` or `consume`? In the example, I invoke `consumer.consume(...)`. I would expect an authentication failure to occur at this invocation given the username is `BROKEN`.

**Comment by edenhill** [comment]
Yes, that's what should be happening.

Suggest you upgrade to the latest version of the client.

**Comment by ncernek** [comment]
I upgraded to `1.6.0`, now I am getting error messages, (good!), but no raised error. This occurs on `consumer.subscribe(...)`
```
%3|1615490869.628|FAIL|rdkafka#consumer-1| [thrd:sasl_ssl://myserver:9092/bootstra]: sasl_ssl://myserver:9092/bootstrap: SASL authentication error: Authentication failed (after 1070ms in state AUTH_REQ)```

How do I raise this error? According to the [docs](https://confluent-kafka-python.readthedocs.io/en/latest/#pythonclient-configuration) `error_cb` is not called for `subscribe`. The [subscribe](https://confluent-kafka-python.readthedocs.io/en/latest/#confluent_kafka.Consumer.subscribe) method itself doesn't have any params for error handlers.

**Comment by edenhill** [comment]
The error_cb is triggered by poll().

Den tors 11 mars 2021 kl 20:31 skrev Nico Cernek ***@***.***>:

**Comment by ncernek** [comment]
Yes. But `poll` is not triggering `error_cb`, and yet I see this error log. So, how do I raise the error?

**Comment by edenhill** [comment]
The logs that you see are from background librdkafka threads, they are unrelated to any API calls you are making.

General errors are propagated through error_cb, which needs to be trigger on the application thread - through poll() - your error_cb will not trigger on another thread.

**Comment by ncernek** [comment]
I see. So there's a bug, as `error_cb` is not being triggered at all.

**Comment by edenhill** [comment]
You have a timeout of 1 so it is possible that the timeout elapses before the connection attempt fails.

**Comment by ncernek** [comment]
I tried this with a timeout of 10 seconds, with the same problem. I see a log that says `auth failed`, but `error_cb` is not triggered.

**Comment by MSeal** [comment]
Closing as this was an old ticket and the async interaction pattern has changed substantially since
